Add a guide on using different types of tracking server (#10241)
mlflow/mlflow: docs/source/getting-started/tracking-server-overview/index.rst

.. _tracking-server-overview:

5 Minute Tracking Server Overview

> ======================================================================

Reviewer: The header length can extend to the length of the preceding line (just for consistency). It's fine if it's longer to sphinx (just not shorter), but my control-freak nature twitches when I see these extend beyond the header text ;)

In this guide we will walk you through how to view your MLflow experiment results with different types of 

> tracking server. At a high level there are 3 ways to view your MLflow experiments:

Reviewer: ```suggestion
tracking server configurations. At a high level, there are 3 ways to view your MLflow experiments:
```

> * **[Method 1]** Start your own MLflow server.

Reviewer: love the bullets

* **[Method 2]** Use a free hosted tracking server - Databricks Community Edition.
* **[Method 3]** Use production Databricks/AzureML.

To choose among these 3 methods, here is our recommendation:

* If you have privacy concerns (data/model/tech stack), use **Method 1 - start your own server**.

> * If you are a student/individual researcher, or if you are developing in cloud-based notebook, e.g., Google 

Reviewer: ```suggestion
* If you are a student or an individual researcher, or if you are developing in cloud-based notebooks (e.g., Google 
```

>   Colab, use **Method 2 - free hosted tracking server**.

Reviewer: ```suggestion
  Colab), use **Method 2 - free hosted tracking server**.
```

> * Enterprise users, or if you want to serve/productionize your model, please use 

Reviewer: ```suggestion
* Enterprise users, or if you want to serve or deploy your model for a production use-case, please use 
```

  **Method 3 - production Databricks/AzureML**.

Overall **Method 2 - free hosted tracking server** is the simplest way to get started with MLflow, but please
pick the method that best suits your needs.

Method 1: Start Your Own MLflow Server
---------------------------------------

> **Disclaimier**: This part of guide is not suitable for running in a cloud-provided IPython environment, e.g., 

Reviewer: can we wrap the exempli gratia statements in parentheses and just use a comma separator between the short list of examples? 
(e.g., Collab, Databricks)

> Google Colab or Databricks Notebook. Please follow the guide below in your local machine (laptop/desktop).

Reviewer: ```suggestion
Google Colab or Databricks Notebook. Please follow the guide below in your local computer.
```

> Hosted tracking server is the simplest way to store and view MLflow experiments, but it is not suitable for 

Reviewer: ```suggestion
A hosted tracking server is the simplest way to store and view MLflow experiments, but it is not suitable for 
```

> every user. For example, you don't want to expose your data and model to cloud provider. In this case, 

Reviewer: ```suggestion
every user. For example, you may not want to expose your data and model to others in your cloud provider account. In this case, 
```
I'm not sure if anyone ever thinks it's risky to share data to a cloud provider; it's more of the security of other people who have access to that account :)

you can use a local hosted MLflow server to store and view your experiments. To do so, there are two steps:

* Start your MLflow server.

> * Port MLflow session to the local MLflow server IP by :py:func:`mlflow.set_tracking_uri()`.

Reviewer: "Port" seems a bit odd. Submit? Send?

Author: Maybe "connect"? basically this API hooks mlflow session together with the tracking server.

Start a Local MLflow Server
^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you don't have MLflow installed, please run the command below to install it:

.. code-block:: bash

    $ pip install mlflow

> Installing MLflow will also install the MLflow CLI tool, you can start a local MLflow server with UI 

Reviewer: ```suggestion
The installation of MLflow includes the MLflow CLI tool, so you can start a local MLflow server with UI 
```

by running the command below in your terminal:

.. code-block:: bash

    $ mlflow ui

It will generate logs with the IP address, for example:

.. code-block:: bash

    (mlflow) [master][~/Documents/mlflow_team/mlflow]$ mlflow ui
    [2023-10-25 19:39:12 -0700] [50239] [INFO] Starting gunicorn 20.1.0
    [2023-10-25 19:39:12 -0700] [50239] [INFO] Listening at: http://127.0.0.1:5000 (50239)

> Open the URL in your browser will bring you to the MLflow UI. This is the open source version MLflow UI, 

Reviewer: ```suggestion
Opening the URL of the MLflow tracking server in your browser will bring you to the MLflow UI. The image below is from the open source version of the MLflow UI, 
```

which is a bit different from the MLflow UI on Databricks CE. Below is a screenshot of the landing page:

.. image:: ../../_static/images/quickstart/tracking-server-overview/mlflow-localhost-landing-page.png
    :width: 800px
    :align: center
    :alt: Landing page of OSS MLflow server

> It's also possible to deploy your own MLflow server on cloud platforms, but it is out of the scope of this guide.

Reviewer: ```suggestion
It's also possible to deploy your own MLflow server on cloud platforms, but that process is a bit more involved and is out of the scope of this particular guide.
```

Reviewer: Can we mark this as a 

```rst

.. note:: 

``` 

element?

Connect MLflow Session to Your Server
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Now the server is spinned up, let's connect our MLflow session to the local server. This is very 

Reviewer: ```suggestion
Now that the server is spun up, let's connect our MLflow session to the local server. This is very 
```

> similar to how we connect to Databricks platform.

Reviewer: ```suggestion
similar to how we connect to a remote hosted tracking provider such as the Databricks platform.
```

.. code-block:: python

  mlflow.set_tracking_uri("http://localhost:5000")

> Then let's try logging some dummy metrics, and view them on the local hosted UI:

Reviewer: ```suggestion
Next, let's try logging some dummy metrics. We can view these test metrics on the local hosted UI:
```

.. code-block:: python

  mlflow.set_experiment("/check-localhost-connection")

  with mlflow.start_run():
      mlflow.log_metric("foo", 1)
      mlflow.log_metric("bar", 2)

> Putting it together you can copy the following code to your editor and save it as `test_mlflow_localhost.py`:

Reviewer: ```suggestion
Putting it all together, you can copy the following code to your editor and save it as `test_mlflow_localhost.py`:
```

Reviewer: can we change the name of the script since files and functions starting with `test` have the connotation of being associated with pytest?

Author: renamed to `log_mlflow_with_localhost.py`

.. code-block:: python

  import mlflow

  mlflow.set_tracking_uri("http://localhost:5000")
  mlflow.set_experiment("check-localhost-connection")

  with mlflow.start_run():
      mlflow.log_metric("foo", 1)
      mlflow.log_metric("bar", 2)

Then execute it by:

.. code-block:: bash

  $ python test_mlflow_localhost.py

View Experiment on Your MLflow Server
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Now let's view your experiment on the local server. Open the URL in your browser, which is `http://localhost:5000` 
in our case. In the UI, inside the left sidebar you should see the experiment with name 

> `"check-localhost-connection"`, clicking on it should bring you to the experiment view similar as below. 

Reviewer: ```suggestion
`"check-localhost-connection"`. Clicking on this experiment name should bring you to the experiment view, similar to what is shown below. 
```

.. image:: ../../_static/images/quickstart/tracking-server-overview/mlflow-localhost-experiment-view.png
    :width: 800px
    :align: center
    :alt: Experiment view of OSS MLflow server

> Clicking on the run (`"clumsy-steed-426"`) will bring you to the run view, similar as below.

Reviewer: ```suggestion
Clicking on the run ( in this example,`"clumsy-steed-426"`, yours will be different!) will bring you to the run view, similar as below.
```

.. image:: ../../_static/images/quickstart/tracking-server-overview/mlflow-localhost-run-view.png
    :width: 800px
    :align: center
    :alt: Run view of OSS MLflow server

Conclusion
^^^^^^^^^^^

That's all about how to start your own MLflow server and view your experiments. Please see the pros and cons
of this method below:

* **Pros**
  
  * You have full control of your data and model, which is good for privacy concerns.
  * No subscription is required.
  * Unlimited quota of experiments/runs.
  * You can even customize your UI by forking the MLflow repo and modify the UI code.

* **Cons**
  
  * Requires manual setup and maintenance.

>   * Team collaboration is harder than using a hosted tracking server, e.g., Databricks CE.

Reviewer: Databricks CE isn't collaborative, though. It's single-user, right? We might not want to push too hard on the "use Databricks" angle here. Can we make this provider agnostic (AzureML has hosted MLflow as well)

Author: done

  * Not suitable for cloud-based notebook, e.g., Google Colab.
  * Requires extra port forwarding if you deploy your server on cloud VM.
  * No serving support.

Method 2: Use Free Hosted Tracking Server (Databricks Community Edition)
------------------------------------------------------------------------

**Notice**: This part of guide can be directly executed in cloud-based notebook, e.g., Google Colab or 
Databricks Notebook.

> `Databricks Community Edition (CE) <https://community.cloud.databricks.com/>`_ is the free version of 

Reviewer: ```suggestion
`Databricks Community Edition (CE) <https://community.cloud.databricks.com/>`_ is the free, limited-use version of the 
```

Reviewer: It definitely has a LOT of restrictions.

cloud-based big data platform Databricks. Databricks CE users can access a micro-cluster as well as 
a cluster manager and notebook environment. All users can share their notebooks and host them free of 
charge with Databricks. You can use Databricks CE to store and view your MLflow experiments without 

> being charged.

Reviewer: We might want to clean this wording up a bit. It's great for getting started and learning the basics of Databricks / MLflow, but it's definitely not intended for long-term use (I think there's a purge that happens, so there's not even a guarantee of data consistency and longevity; we definitely don't want to make it seem like people can use CE as a solution in and of itself. It's just for teaching / learning purposes only.

Author: CE is actually okay for long-term use if users just want to track, which covers a decent portion of users. I did some search and the experiments won't be deleted automatically. My experiments from 3 months ago are still accessible.

From next Q2 or so, we can refactor this part to rely on Lighthouse.

To use Databricks CE to store and view our MLflow experiments, basically we need to:

> 

Reviewer: * Create a free Databricks CE account

* Set up Databricks CE authentication in our dev environment.
* Connect to Databricks CE in our MLflow experiment session.

Then the experiment results will be automatically sent to Databricks CE, where you can view it in 
MLflow experiment UI. Now let's look at the code.

Create a Databricks CE Account
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you don't have an account of Databricks CE yet, you can create one 

> `here <https://www.databricks.com/try-databricks#account>`_, it should take no longer than 3 minutes.

Reviewer: ```suggestion
`here <https://www.databricks.com/try-databricks#account>`_. The full process should take no longer than 3 minutes.
```

.. image:: ../../_static/images/quickstart/tracking-server-overview/databricks-ce-experiment-view.png
    :width: 800px
    :align: center
    :alt: Experiment view of Databricks MLflow server

Clicking on the run name, in our example it is "youthful-lamb-287" (it's a randomly generated name, you will see 
a different name in your CE console), will bring you to the run view, similar to 

.. image:: ../../_static/images/quickstart/tracking-server-overview/databricks-ce-run-view.png
    :width: 800px
    :align: center
    :alt: Experiment view of Databricks MLflow server

In the run view, you will see our dummy metrics `"foo"` and `"bar"` are logged successfully.

Conclusion
^^^^^^^^^^^

That's all about how to use Databricks CE as the tracking server. Please see the pros and cons
of this method below:

* **Pros**
  
  * Effortless setup.
  * Free. 
  * Good for collaboration, e.g., you can share your MLflow experiment with your teammates easily.
  * Compatible for developing on cloud-based notebook, e.g., Google Colab.
  * Compatible for developing on cloud VM.

* **Cons**
  
  * Has quota limit of experiments/runs.
  * No model registration/serving support.

> Method 3: Use Production Databricks Workspace

Reviewer: Can we make this provider agnostic? SaaS MLflow Workspace and give an exempli gratia list inclusive of Databricks, AzureML

Author: done
